packages/edit-widgets/src/components/header/index.js

```diff
@@ -2,75 +2,96 @@
 
 function Header() {
+	const inserterButton = useRef();
 	const isLargeViewport = useViewportMatch( 'medium' );
-	const rootClientId = useLastSelectedRootId();
 	const isLastSelectedWidgetAreaOpen = useSelect(
 		( select ) =>
 			select( 'core/edit-widgets' ).getIsWidgetAreaOpen( rootClientId ),
 		[ rootClientId ]
 	);
-	const { setIsWidgetAreaOpen } = useDispatch( 'core/edit-widgets' );
+	const isInserterOpened = useSelect( ( select ) =>
+		select( 'core/edit-widgets' ).isInserterOpened()
+	);
+	const { setIsWidgetAreaOpen, setIsInserterOpened } = useDispatch(
+		'core/edit-widgets'
+	);
 	const { selectBlock } = useDispatch( 'core/block-editor' );
-
-	function handleInserterOpen( isOpen ) {
-		if ( isOpen && ! isLastSelectedWidgetAreaOpen ) {
-			// Select the last selected block if hasn't already.
-			selectBlock( rootClientId );
-			// Open the last selected widget area when opening the inserter.
-			setIsWidgetAreaOpen( rootClientId, isOpen );
-		}
-	}
+	const rootClientId = useLastSelectedRootId();
 
 	return (
 		<>
 			<div className="edit-widgets-header">
-				<NavigableToolbar
-					className="edit-widgets-header-toolbar"
-					aria-label={ __( 'Document tools' ) }
-				>
-					<ToolbarItem>
-						{ ( toolbarItemProps ) => (
-							<Inserter
-								position="bottom right"
-								showInserterHelpPanel
-								toggleProps={ {
-									...inserterToggleProps,
-									...toolbarItemProps,
-								} }
-								rootClientId={ rootClientId }
-								onToggle={ handleInserterOpen }
-							/>
-						) }
-					</ToolbarItem>
-					<ToolbarItem as={ UndoButton } />
-					<ToolbarItem as={ RedoButton } />
-					<ToolbarItem as={ BlockNavigationDropdown } />
-				</NavigableToolbar>
-				<h1 className="edit-widgets-header__title">
-					{ __( 'Block Areas' ) }
-				</h1>
+				<div className="edit-widgets-header__navigable-toolbar-wrapper">
+					<h1 className="edit-widgets-header__title">
+						{ __( 'Widgets' ) }
+					</h1>
+					<NavigableToolbar
+						className="edit-widgets-header-toolbar"
+						aria-label={ __( 'Document tools' ) }
+					>
+						<ToolbarItem
+							ref={ inserterButton }
+							as={ Button }
+							className="edit-widgets-header-toolbar__inserter-toggle"
+							isPrimary
+							isPressed={ isInserterOpened }
+							onMouseDown={ ( event ) => {
+								event.preventDefault();
+							} }
+							onClick={ () => {
+								if ( isInserterOpened ) {
+									// Focusing the inserter button closes the inserter popover
+									inserterButton.current.focus();
+								} else {
+									if ( ! isLastSelectedWidgetAreaOpen ) {
+										// Select the last selected block if hasn't already.
+										selectBlock( rootClientId );
+										// Open the last selected widget area when opening the inserter.
+										setIsWidgetAreaOpen(
+											rootClientId,
+											true
+										);
+									}
+									// eslint-disable-next-line @wordpress/react-no-unsafe-timeout
```

Comment:
Why do we need this?

Reply:
Without it, eslint complains about doing setTimeout without matching clearTimeout. In this case the setTimeout call is all about running code on the next tick so clearTimeout wouldn't make as much sense.

Reply:
Ahh I meant to comment the next line: why do we need the `setTimeout`?

Reply:
Without it, the DOM updates resulting from `selectBlock()` and `setIsInserterOpened()` call are applied the same tick and pretty much in a random order. The inserter is closed if any other part of the app gets focus. If `selectBlock()` happens to take effect after `setIsInserterOpened()` then the inserter is open for a brief moment and then gets auto-closed due to focus moving to selected block.

Reply:
Maybe `requestAnimationFrame` would be a better choice? And would also appreciate if you can add some comments above it, since it's not idiomatic code 😅